File: 2017s/soln/ImageLock.java
import java.util.*;
import java.io.*;
public class ImageLock {
    public static void main(String[] args) throws FileNotFoundException {
        // Before submitting, make sure the main method hasn't been changed!
        Scanner sc = new Scanner(new FileReader("ImageLockIN.txt"));
        while(sc.hasNext()) {
            int width = sc.nextInt();
            int height = sc.nextInt();
            int[][] image = new int[width][height];
            for (int i = 0; i < height; i++) {
              for (int j = 0; j < width; j++) {
                image[j][i] = sc.nextInt();
              }
            }
            if (isCat(image)) {
              System.out.println("cat");
            } else {
              System.out.println("random");
            }
        }
        sc.close();
    }
    
    public static boolean isCat(int[][] image) {
        boolean ans = true;
        //STUBIFY
        int adj = 0;
        double BOUND = 120.0;
        for (int i = 0; i < image.length; i++) {
            for (int j = 0; j < image[0].length - 1; j++) {
              adj += Math.abs(image[i][j] - image[i][j + 1]);
            }
        }
        for (int i = 0; i < image.length - 1; i++) {
            for (int j = 0; j < image[0].length; j++) {
              adj += Math.abs(image[i][j] - image[i + 1][j]);
            }
        }
        int dim = image.length * image[0].length;
        double ratio = (double)adj / (double)dim;
        //ENDSTUBIFY
        return ratio < BOUND;
    }
}

File: 2017s/soln/ImageLock.py
def isCat(img):
    numagrees = 0
    W = len(img)
    H = len(img[0])
    for x in range(1,(W-1)//2):
        for y in range(1,(H-1)//2):
            c = 0
            for (i, j) in [(-1,-1),(-1,0),(-1,1),(0,-1),(0,1),(1,-1),(1,0),(1,1)]:
                if abs(img[x][y] - img[x+i][y+j]) < 5:
                    c = c + 1
            if c >= 4:
                numagrees = numagrees + 1
    return numagrees/float(W*H//4) > 0.01 




if __name__ == "__main__":
    with open("ImageLockIN.txt", "r") as f:
        while True:
            s = f.readline()
            if s == "":
                break
            s = s.split(" ")
            W = int(s[0])
            H = int(s[1])
            img = [ [ 0 for i in range(H) ] for j in range(W) ]
            for y in range(H):
                s = f.readline()
                s = s.split(" ")
                for x in range(W):
                    img[x][y] = int(s[x])

            if isCat(img):
                print('cat')
            else:
                print('random')



